pkg/services/ngalert/state/cache.go

```diff
@@ -261,3 +261,9 @@ func mergeLabels(a, b data.Labels) data.Labels {
 	}
 	return newLbs
 }
+
+func (c *cache) deleteEntry(orgID int64, alertRuleUID, cacheID string) {
+	c.mtxStates.Lock()
+	defer c.mtxStates.Unlock()
```

Comment:
Given that we call this in a loop, I wonder if it might be better to avoid the unlock defer in these when they are short and also don't have any way of returning an error before the unlock. I think it adds a small amount of overhead. No strong feelings on this, but it keeps popping into my head so wanted to get it out :-)

Reply:
I could go for that for as long as the only thing happening in here is a map delete, which is just a noop when the key isn't present (or the nested map is nil/doesn't exist). If we were to start doing anything else in this method then we would need to reconsider that, and having it here makes it less likely we'll forget to add it back and possibly introduce a race condition

Reply:
Since there are multiple candidates (e.g. cache.reset()) , probably best to leave it out of this PR, and if we want to do something like that, do it another PR as a group.